src/xdp/xsk.c

```diff
@@ -960,6 +990,51 @@ XskDereferenceDatapathHandle(
     XskDereference(Xsk);
 }
 
+_IRQL_requires_max_(PASSIVE_LEVEL)
+_IRQL_requires_same_
+NTSTATUS
+XskCreate(
+    _Out_ XSK **Xsk,
+    _In_opt_ PXSK_NOTIFY_CALLBACK NotifyCallback
+    )
+{
+    NTSTATUS Status = STATUS_SUCCESS;
+
+    TraceEnter(TRACE_XSK, "Xsk=%p", Xsk);
+
+    *Xsk = ExAllocatePoolZero(NonPagedPoolNx, sizeof(**Xsk), POOLTAG_XSK);
+    if (*Xsk == NULL) {
+        Status = STATUS_INSUFFICIENT_RESOURCES;
+        goto Exit;
+    }
+
+    (*Xsk)->Header.ObjectType = XDP_OBJECT_TYPE_XSK;
+    (*Xsk)->Header.Dispatch = &XskFileDispatch;
+    XdpInitializeReferenceCount(&(*Xsk)->ReferenceCount);
+    (*Xsk)->State = XskUnbound;
+    (*Xsk)->Rx.Xdp.HookId.Layer = XDP_HOOK_L2;
+    (*Xsk)->Rx.Xdp.HookId.Direction = XDP_HOOK_RX;
+    (*Xsk)->Rx.Xdp.HookId.SubLayer = XDP_HOOK_INSPECT;
+    (*Xsk)->Tx.Xdp.HookId.Layer = XDP_HOOK_L2;
+    (*Xsk)->Tx.Xdp.HookId.Direction = XDP_HOOK_TX;
+    (*Xsk)->Tx.Xdp.HookId.SubLayer = XDP_HOOK_INJECT;
+    KeInitializeSpinLock(&(*Xsk)->Lock);
+    KeInitializeEvent(&(*Xsk)->IoWaitEvent, NotificationEvent, TRUE);
+    KeInitializeEvent(&(*Xsk)->PollRequested, SynchronizationEvent, FALSE);
+    KeInitializeEvent(&(*Xsk)->Tx.Xdp.OutstandingFlushComplete, NotificationEvent, FALSE);
+    (*Xsk)->NotifyCallback = NotifyCallback;
+
+    EventWriteXskCreateSocket(
+        &MICROSOFT_XDP_PROVIDER, *Xsk, PsGetCurrentProcessId());
```

Comment:
Doesn't ETW already capture current process ID if called in the context of a user mode process?

Reply:
Yes, it captures process, thread, and CPU.

Reply:
Then why is `PsGetCurrentProcessId()` an explicit argument?

Reply:
It's there for two scenarios, actually: 1. **W**SK allows sockets to be created on behalf of specific processes (i.e. not the current process) and 2. we want to be able to execute this code in arbitrary contexts, not necessarily handle this code inline within the IRP request.

So we want consumers of this event to take a dependency on the PID we supply, not the ETW PID.

Reply:
"WSK allows sockets to be created on behalf of specific processes (i.e. not the current process) which we wanted to be able to support for kernel AF_XDP too", I was unaware of this requirement, but I can adjust the signature of the kernel-mode XskCreate no problem since it is already going to be different than the user-mode one.

Reply:
Updated the KM XskCreate fn definition